src/Changelog/ReleaseChangelogViaKeepAChangelog.php

```diff
@@ -0,0 +1,141 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Laminas\AutomaticReleases\Changelog;
+
+use Laminas\AutomaticReleases\Git\CommitFile;
+use Laminas\AutomaticReleases\Git\Push;
+use Laminas\AutomaticReleases\Git\Value\BranchName;
+use Laminas\AutomaticReleases\Git\Value\SemVerVersion;
+use Phly\KeepAChangelog\Common\DiscoverChangelogEntryListener;
+use Phly\KeepAChangelog\Config;
+use Phly\KeepAChangelog\Version\ReadyLatestChangelogEvent;
+use Phly\KeepAChangelog\Version\SetDateForChangelogReleaseListener;
+use Symfony\Component\Console\Output\NullOutput;
+
+use function date;
+use function file_exists;
+use function sprintf;
+
+final class ReleaseChangelogViaKeepAChangelog implements ReleaseChangelog
+{
+    private CommitFile $commitFile;
+    private Push $push;
+
+    public function __construct(
+        CommitFile $commitFile,
+        Push $push
+    ) {
+        $this->commitFile = $commitFile;
+        $this->push       = $push;
+    }
+
+    /**
+     * @psalm-param non-empty-string $repositoryDirectory
+     */
+    public function __invoke(
+        string $repositoryDirectory,
+        SemVerVersion $version,
+        BranchName $sourceBranch
+    ): void {
+        $changelogFile = $repositoryDirectory . '/CHANGELOG.md';
+        if (! file_exists($changelogFile)) {
+            // No changelog
+            return;
+        }
+
+        $versionString = $version->fullReleaseName();
+
+        if (! $this->updateChangelog($changelogFile, $versionString)) {
+            // Failure to update; nothing to commit
+            return;
+        }
+
+        ($this->commitFile)(
+            $repositoryDirectory,
+            $sourceBranch,
+            'CHANGELOG.md',
+            sprintf('%s readiness', $versionString)
+        );
+
+        ($this->push)($repositoryDirectory, $sourceBranch->name());
+    }
+
+    private function updateChangelog(string $changelogFile, string $versionString): bool
+    {
+        $event = $this->createReadyLatestChangelogEvent($changelogFile, $versionString);
+
+        (new DiscoverChangelogEntryListener())($event);
```

Comment:
These listeners are bizarre: I know they are mostly imported, and you mostly did C&P of some upstream code: still feels weird for them to mutate `$event`.

Probably design similar to what we had in `laminas/laminas-mvc`?

Reply:
PSR-14, actually, via the `StoppableEventInterface`. Essentially, dispatchers check for `$event->isPropagationStopped()` after triggering each listener; if it returns true, it doesn't iterate to the next listener. I use this heavily in keep-a-changelog, as it makes it easier for me to report error conditions back to the user and/or to stop doing work if nothing else needs to be done. I use a `failed()` method to check for it in my library, and `isPropagationStopped()` typically checks for that (sometimes it checks other things as well).